// The definition of mayastor gRPC interface.
// This interface is supposed to be independent on particular computing
// environment (i.e. kubernetes).

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.openebs.mayastor";
option java_outer_classname = "MayastorProto";

package mayastor;

// Service for managing storage pools, replicas and nexus's.
// Replica holds the actual user data. From user application the
// data are accessed indirectly through nexus object.
service Mayastor {
	// Storage pool related methods.
	//
	// Storage pool is made up of block devices disks and provides a storage
	// space for thin provisioning of replicas.

	rpc CreatePool (CreatePoolRequest) returns (Null) {}
	rpc DestroyPool (DestroyPoolRequest) returns (Null) {}
	rpc ListPools (Null) returns (ListPoolsReply) {}

	// Replica related methods.
	//
	// Replica allocates space from storage pool.

	rpc CreateReplica (CreateReplicaRequest) returns (CreateReplicaReply) {}
	rpc DestroyReplica (DestroyReplicaRequest) returns (Null) {}
	rpc ListReplicas (Null) returns (ListReplicasReply) {}
	rpc StatReplicas (Null) returns (StatReplicasReply) {}
	rpc ShareReplica (ShareReplicaRequest) returns (ShareReplicaReply) {}

	// Nexus related methods.
	//
	// Nexus is a logical frontend representing a data volume taking care of
	// replication and rebuild in the background.

	rpc CreateNexus (CreateNexusRequest) returns (Null){}
	rpc DestroyNexus (DestroyNexusRequest) returns (Null) {}
	rpc ListNexus (Null) returns (ListNexusReply) {}
	rpc AddChildNexus (AddChildNexusRequest) returns (Null) {}
	rpc RemoveChildNexus (RemoveChildNexusRequest) returns (Null) {}

	// This method is called by control plane to construct a block device
	// (/dev/...) that will be used to connect the nexus to the OS.
	rpc PublishNexus (PublishNexusRequest) returns (PublishNexusReply) {}
	rpc UnpublishNexus (UnpublishNexusRequest) returns (Null) {}

	// Nexus child operations
	rpc ChildOperation(ChildNexusRequest) returns (Null) {}

	// Rebuild operations
	rpc StartRebuild (StartRebuildRequest) returns (Null) {}
	rpc StopRebuild (StopRebuildRequest) returns (Null) {}
	rpc PauseRebuild (PauseRebuildRequest) returns (Null) {}
	rpc ResumeRebuild (ResumeRebuildRequest) returns (Null) {}
	rpc GetRebuildState (RebuildStateRequest) returns (RebuildStateReply) {}
	rpc GetRebuildProgress (RebuildProgressRequest) returns (RebuildProgressReply) {}
}

// Means no arguments or no return value.
message Null {}

// I/O interface used for underlying disks in a pool
enum PoolIoIf {
  POOL_IO_AUTO = 0;    // prefer uring if supported, falling back to aio
  POOL_IO_AIO = 1;     // Linux AIO
  POOL_IO_URING = 2;   // io_uring, requires Linux 5.1
}

// Create pool arguments.
// Currently we support only concatenation of disks (RAID-0).
message CreatePoolRequest {
  string name = 1;           // name of the pool
  repeated string disks = 2; // disk device paths or URIs to be claimed by the pool
  uint32 block_size = 3; // when using files, we need to specify the block_size
  PoolIoIf io_if = 4;        // I/O interface
}

// State of the storage pool (terminology comes from ZFS).
enum PoolState {
  POOL_UNKNOWN = 0;
  POOL_ONLINE = 1;   // the pool is in normal working order
  POOL_DEGRADED = 2; // the pool has experienced a failure but can still function
  POOL_FAULTED = 3;  // the pool is completely inaccessible
}

// Storage pool properties
message Pool {
  string name = 1;            // name of the pool
  repeated string disks = 2;  // absolute disk paths claimed by the pool
  PoolState state = 3;        // current state of the pool
  uint64 capacity = 5;        // size of the pool in bytes
  uint64 used = 6;            // used bytes from the pool
}

// Destroy pool arguments.
message DestroyPoolRequest {
  string name = 1;  // name of the pool
}

// List of pools and their properties.
message ListPoolsReply {
  repeated Pool pools = 1;  // list of the pools
}

// Protocol for remote storage access which exposes a replica.
enum ShareProtocolReplica {
  REPLICA_NONE = 0;   // not exposed
  REPLICA_NVMF = 1;   // NVMe over Fabrics (TCP)
  REPLICA_ISCSI = 2;  // iSCSI
}

// Note that enum values use C++ scoping rules, meaning that enum values are siblings of their type,
// not children of it.
// So cannot use NBD, NVMF, and ISCSI as symbols for ShareProtocolNexus
enum ShareProtocolNexus {
  NEXUS_NBD = 0;    // local
  NEXUS_NVMF = 1;   // NVMe over Fabrics (TCP)
  NEXUS_ISCSI = 2;  // iSCSI
}

// Create replica arguments.
message CreateReplicaRequest {
  string uuid = 1;  // uuid of the replica
  string pool = 2;  // name of the pool
  uint64 size = 3;  // size of the replica in bytes
  bool thin = 4;    // thin provisioning
  ShareProtocolReplica share = 5;  // protocol to expose the replica over
}

// Create replica response.
message CreateReplicaReply {
  string uri = 1;   // uri under which the replica is accessible by nexus
}

// Destroy replica arguments.
message DestroyReplicaRequest {
  string uuid = 1;  // name of the replica
}

// Replica properties
message Replica {
  string uuid = 1;  // uuid of the replica
  string pool = 2;  // name of the pool
  bool thin = 3;    // thin provisioning
  uint64 size = 4;  // size of the replica in bytes
  ShareProtocolReplica share = 5;  // protocol used for exposing the replica
  string uri = 6;   // uri usable by nexus to access it
}

// List of replicas and their properties.
message ListReplicasReply {
  repeated Replica replicas = 1;  // list of the replicas
}

// NOTE: We use struct instead of more suitable map type, because JS protobuf
// lib has problem (yields garbage) when decoding maps containing u64:
// https://github.com/protobufjs/protobuf.js/issues/1203
message Stats {
  uint64 num_read_ops = 1;
  uint64 num_write_ops = 2;
  uint64 bytes_read = 3;
  uint64 bytes_written = 4;
}

// Replica stats
message ReplicaStats {
  string uuid = 1;  // uuid of the replica
  string pool = 2;  // name of the pool
  Stats stats = 3;  // stat counters
}

// List of replicas and their properties.
message StatReplicasReply {
  repeated ReplicaStats replicas = 1;  // list of the replicas
}

// Share replica request.
message ShareReplicaRequest {
  string uuid = 1;  // uuid of the replica
  ShareProtocolReplica share = 2;  // protocol used for exposing the replica
                            // Use "NONE" to disable remote access.
}

// Share replica response.
message ShareReplicaReply {
  string uri = 1;   // uri under which the replica is accessible by nexus
}

// Create nexus arguments.
message CreateNexusRequest {
  string uuid = 1; // this UUID will be set in as the UUID
  uint64 size = 2; // size of the device in bytes
  // replica can be iscsi and nvmf remote targets or a local spdk bdev
  // (i.e. bdev:///name-of-the-bdev).
  repeated string children = 3; // uris to the targets we connect to
}

// State of the nexus child.
enum ChildState {
  CHILD_UNKNOWN = 0;
  CHILD_ONLINE = 1;   // healthy and contains the latest bits
  CHILD_DEGRADED = 2; // rebuild is in progress (or other recoverable error)
  CHILD_FAULTED = 3;  // unrecoverable error (control plane must act)
}

// represents a child device part of a nexus
message Child {
  string uri = 1;   // uri of the child device
  ChildState state = 2; // state of the child
  int32 rebuild_progress = 3;
}

// State of the nexus (terminology inspired by ZFS).
enum NexusState {
  NEXUS_UNKNOWN = 0;
  NEXUS_ONLINE = 1;    // healthy and working
  NEXUS_DEGRADED = 2;  // not healthy but is able to serve IO (i.e. rebuild is in progress)
  NEXUS_FAULTED = 3;   // broken and unable to serve IO
}

// represents a nexus device
message Nexus {
  string uuid = 1;             // name of the nexus
  uint64 size = 2;             // size of the volume in bytes
  NexusState state = 3;        // current state of the nexus
  repeated Child children = 4; // array of children
  // Path to device file for the volume (missing if not published).
  // Missing property and empty string are treated the same.
  string device_path = 5;
  uint32 rebuilds = 6;         // total number of rebuild tasks
}

message ListNexusReply {
  repeated Nexus nexus_list = 1;
}

message DestroyNexusRequest   {
  string uuid = 1;    // uuid of the nexus
}

message AddChildNexusRequest {
  string uuid = 1;    // uuid of the nexus
  string uri = 2;     // URI of the child device to be added
  bool norebuild = 3;   // auto start rebuilding
}

message RemoveChildNexusRequest {
  string uuid = 1;    // uuid of the nexus
  string uri = 2;     // URI of the child device to be removed
}

// this message will be subject to change as we will add support for remote
// storage protocols.
message PublishNexusRequest {
  string uuid = 1; // uuid of the nexus which to create device for
  string key = 2; // encryption key
  ShareProtocolNexus share = 3;  // protocol used for the front end.
}

message PublishNexusReply {
  string device_path = 1; // i.e. /dev/nbdX
}

message UnpublishNexusRequest {
  string uuid = 1;   // uuid of the nexus which to destroy
}

enum ChildAction {
  offline= 0;
  online = 1;
}

message ChildNexusRequest {
  string uuid = 1;
  string uri =2;
  ChildAction action = 3;
}

message RebuildStateRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the destination child
}

message RebuildStateReply {
  string state = 1; // current rebuild state (i.e. ready/running/completed etc.)
}

message StartRebuildRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the child to be rebuilt
}

message StopRebuildRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the destination child
}

message PauseRebuildRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the destination child
}

message ResumeRebuildRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the destination child
}

message RebuildProgressRequest {
  string uuid = 1;  // uuid of the nexus
  string uri = 2;   // uri of the destination child
}

message RebuildProgressReply {
  uint32 progress = 1;  // progress percentage
}


// Anything what follows here are private interfaces used for testing or
// debugging mayastor. Do not use it for anything serious.

service BdevRpc {
	rpc List(mayastor.Null) returns (Bdevs) {}
	rpc Create(BdevUri) returns (CreateReply) {}
}

message Bdev {
	string name = 1;
	string uuid = 2;
	uint64 num_blocks = 3;
	uint32 blk_size = 5;
	bool claimed = 6;
	string claimed_by = 7;
}

message Bdevs {
	repeated Bdev bdevs = 1;
}

message BdevUri {
	string uri = 1;
}

message CreateReply {
	string name = 1;
}
